charms/istio-pilot/lib/charms/istio_pilot/v0/istio_gateway_name.py

```diff
@@ -0,0 +1,175 @@
+            raise GatewayRelationMissingError()
+        if len(gateway) > 1:
+            raise GatewayRelationTooManyError()
+
+        remote_app = [
+            app
+            for app in gateway[0].data.keys()
+            if isinstance(app, Application) and not app._is_our_app
+        ][0]
+
+        data = gateway[0].data[remote_app]
+
+        if not "gateway_name" in data:
+            logger.error(
+                "Missing gateway name in gateway relation data. Waiting for gateway creation in istio-pilot"
+            )
+            raise GatewayRelationDataMissingError(
+                "Missing gateway name in gateway relation data. Waiting for gateway creation in istio-pilot"
+            )
+
+        if not "gateway_namespace" in data:
+            logger.error("Missing gateway namespace in gateway relation data")
+            raise GatewayRelationDataMissingError(
+                "Missing gateway namespace in gateway relation data"
+            )
+
+        return {
+            "gateway_name": data["gateway_name"],
+            "gateway_namespace": data["gateway_namespace"],
+        }
+
+
+class GatewayProvider(Object):
+    def __init__(self, charm):
+        super().__init__(charm, DEFAULT_RELATION_NAME)
+        self.charm = charm
+        self.lightkube_client = Client(namespace=self.charm.model.name, field_manager="lightkube")
+        self.gateway_class = create_namespaced_resource(
+            group="networking.istio.io",
+            version="v1beta1",
+            kind="Gateway",
+            plural="gateways",
+            verbs=None,
+        )
+        self.framework.observe(
+            charm.on[DEFAULT_RELATION_NAME].relation_changed, self._on_gateway_relation_changed
+        )
+        self.framework.observe(charm.on.config_changed, self._on_gateway_config_changed)
+        self.framework.observe(charm.on.update_status, self._on_gateway_config_changed)
+
+    def _validate_gateway_exists(self):
+        try:
+            response = self.lightkube_client.get(
+                self.gateway_class, self.model.config['default-gateway'], namespace=self.model.name
+            )
+            return True
+        except ApiError as error:
+            logger.error(str(error))
+
+        return False
+
+    def _on_gateway_relation_changed(self, event):
+        if self.model.unit.is_leader():
+            relations = self.model.relations["gateway"]
+            for relation in relations:
+                relation.data[self.charm.app].update(
+                    {
+                        "gateway_name": self.model.config["default-gateway"]
+                        if self._validate_gateway_exists()
+                        else "",
+                        "gateway_namespace": self.model.name,
+                    }
+                )
+
+    def _on_gateway_config_changed(self, event):
+        if len(self.model.relations[DEFAULT_RELATION_NAME]) > 0:
+            self._on_gateway_relation_changed(event)
```

Comment:
I don't know if we need this.  I think any event triggering `_on_gateway_config_changed` could also trigger `_on_gateway_relation_changed`, or at least directly trigger it without additional logic here (since `_on_gateway_relation_changed` can handle cases with zero relations)